src/core/providers/ogr/qgsogrprovider.cpp

```diff
@@ -1189,6 +1189,17 @@ void QgsOgrProvider::loadFields()
     mPrimaryKeyAttrs << 0;
   }
 
+#if GDAL_VERSION_NUM >= GDAL_COMPUTE_VERSION(3,3,0)
+  // needed for field domain retrieval on GDAL 3.3+
+#if QT_VERSION < QT_VERSION_CHECK(5, 14, 0)
+  QMutex *datasetMutex = nullptr;
+#else
+  QRecursiveMutex *datasetMutex = nullptr;
```

Comment:
you need a recursive mutex because we call OGR wrapped method using a mutex too ? But then for qt < 5.14 the standard mutex will deadlock.
So either you need to wrap the new GDAL/OGR methods in a method that acquires a mutex, or you need to restrict the new code path to GDAL >= 3.3 and QT >= 5.14

Reply:
Actually for Qt < 5.14 we are also always using a recursive mutex for the various get*andMutex methods -- it's just that in Qt < 5.14 you had to use a QMutex with the `QMutex::Recursive` flag set on it in the QMutex constructor. (see qgsogrprovider.h:412). That was deprecated in 5.14 and recursive mutexes are now their own class.

Reply:
> That was deprecated in 5.14 and recursive mutexes are now their own class.

ok, makes sense!